File: app/src/main/java/com/donald/musictheoryapp/Music/ScoreView/SLedgerLines.java
package com.donald.musictheoryapp.Music.ScoreView;

import android.graphics.Canvas;
import android.util.Log;

import com.donald.musictheoryapp.Music.MusicXML.Clef;
import com.donald.musictheoryapp.Music.MusicXML.Note;

public class SLedgerLines extends ScoreStamp
{
	private int numberOfLines;
	private boolean aboveStaff;
	private boolean draws;

	protected SLedgerLines(ScoreView view)
	{
		super(view);
	}

	protected void setNote(Note note, Clef clef)
	{
		int staffPosition = Clef.noteStaffPosition(note, clef);
		if(staffPosition > ScoreView.TOP_LINE_STAFF_BOS + 1)
		{
			aboveStaff = true;
			numberOfLines = (staffPosition - ScoreView.TOP_LINE_STAFF_BOS) / 2;
			draws = true;
		}
		else if(staffPosition < ScoreView.BASE_LINE_STAFF_POS - 1)
		{
			aboveStaff = false;
			numberOfLines = (ScoreView.BASE_LINE_STAFF_POS - staffPosition) / 2;
			draws = true;
		}
		else
		{
			draws = false;
		}
	}

	@Override
	protected void onDraw(Canvas canvas, float posX, float posY)
	{
		if(draws)
		{
			float lineWidth = noteWidth() * 1.8f;
			float factor = aboveStaff ? -1 : 1;
			float currentPosY = (aboveStaff ? -staffHeight() : 0f) + posY
				+ staffStepHeight() * 2 * factor;

			for(int i = 0; i < numberOfLines; i++)
			{
				canvas.drawLine(
					posX - lineWidth / 2, currentPosY,
					posX + lineWidth / 2, currentPosY,
					linePaint()
				);
				currentPosY += staffStepHeight() * 2 * factor;
			}
		}
	}
}

File: app/src/main/java/com/donald/musictheoryapp/music/scoreview/SLedgerLines.kt
package com.donald.musictheoryapp.music.scoreview

import android.graphics.Canvas
import com.donald.musictheoryapp.music.MusicXML.Clef
import com.donald.musictheoryapp.music.MusicXML.Clef.Companion.noteStaffPosition
import com.donald.musictheoryapp.music.MusicXML.Note

internal class SLedgerLines(view: ScoreView) : ScoreStamp(view) {

    var note: Note? = null
    var clef: Clef? = null
    private var numberOfLines = 0
    private var aboveStaff = false
    private var draws = false
    override val relContentWidth = REL_LEDGER_LINE_WIDTH

    fun set(note: Note, clef: Clef) {
        val staffPosition = noteStaffPosition(note, clef)
        if (staffPosition > TOP_LINE_STAFF_BOS + 1) {
            aboveStaff = true
            numberOfLines = (staffPosition - TOP_LINE_STAFF_BOS) / 2
            draws = true
        } else if (staffPosition < BASE_LINE_STAFF_POS - 1) {
            aboveStaff = false
            numberOfLines = (BASE_LINE_STAFF_POS - staffPosition) / 2
            draws = true
        } else {
            draws = false
        }
    }

    override fun onDraw(canvas: Canvas, posX: Float, posY: Float) {
        //val note = note ?: return
        //val clef = clef ?: return
        //set(note, clef)
        if (draws) {
            val factor: Float = if (aboveStaff) (-1).toFloat() else 1.toFloat()
            var currentPosY = ((if (aboveStaff) 0F else staffHeight) + posY + staffStepHeight * 2 * factor)
            for (i in 0 until numberOfLines) {
                // TODO: DRAWING ALIGNMENT IS CHANGED, UPDATE THIS
                canvas.drawLine(
                    posX - width / 2, currentPosY,
                    posX + width / 2, currentPosY,
                    linePaint
                )
                currentPosY += staffStepHeight * 2 * factor
            }
        }
    }
}